src/utils/generator.js

```diff
@@ -137,17 +141,33 @@ export class TemporalBuildNodeEntry extends GeneratorEntry {
     }
   }
 
-  declared: void | AbstractValue | ConcreteValue;
+  declared: void | Value;
   args: Array<Value>;
   // If we're just trying to add roots for the serializer to notice, we don't need a buildNode.
   buildNode: void | GeneratorBuildNodeFunction;
   dependencies: void | Array<Generator>;
   isPure: void | boolean;
   mutatesOnly: void | Array<Value>;
+  optimizationFn:
+    | void
+    | ((VisitEntryCallbacks, Generator, TemporalBuildNodeEntry) => TemporalBuildNodeEntryOptimizationStatus);
 
   visit(callbacks: VisitEntryCallbacks, containingGenerator: Generator): boolean {
     let omit = this.isPure && this.declared && callbacks.canOmit(this.declared);
 
+    if (this.optimizationFn !== undefined && this.declared !== undefined) {
+      let optimizationStatus = this.optimizationFn(callbacks, containingGenerator, this);
```

Comment:
Why do any work at all if `omit` might be true, i.e. it is not clear whether this entire generator entry is going to be needed? I suggest to move the entire optimization down to where `recordDeclaration` happens.

Reply:
The the argument rewrite would happen next to the CSE code (with ugly state mutation), and related concepts are closer together...

Reply:
I've re-worked this. There is no argument re-writing going on here though. We don't mutate any entries anymore, we create new ones instead.